lib/Sema/CSSimplify.cpp

```diff
@@ -11499,6 +11499,67 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyFixConstraint(
       }
     }
 
+    auto anchor = locator->getAnchor();
+    auto *AE = getAsExpr<ApplyExpr>(anchor);
+    if (AE && getContextualTypePurpose(anchor) == CTP_ReturnSingleExpr) {
+      auto resultTy = getType(AE);
+      auto overload = findSelectedOverloadFor(AE->getSemanticFn());
+      auto resultTypeVar = resultTy->getAs<TypeVariableType>();
+      auto applyFnType = overload->openedType->castTo<FunctionType>();
+      // When a function result being returned is a generic parameter that can
+      // be inferred by a closure argument result e.g.
+      //   func callit<T>(_ f: () -> T) -> T {
+      //      f()
+      //   }
+      //
+      //   func caller() -> Int {
+      //      callit {
+      //        print("hello")
+      //      }
+      //   }
+      // let's score a contextual mismatch for the result call a bit higher
+      // to avoid ambiguity because we may have contextual mismatches for
+      // the closure as well.
+      if (overload && resultTypeVar &&
+          resultTypeVar->getImpl().getGenericParameter()) {
+        auto typeInvolvesTypeVar = [](Type type, TypeVariableType *typeVar) {
+          bool contains = false;
+          type.visit([&](Type ty) {
+            if (ty->isEqual(typeVar))
+              contains = true;
+          });
+          return contains;
+        };
+
+        // A single closure argument.
+        if (isExpr<ClosureExpr>(AE->getArg()->getSemanticsProvidingExpr())) {
+          auto param = applyFnType->getParams().front();
+          auto closureParamType =
+              param.getParameterType()->getAs<FunctionType>();
+          if (closureParamType &&
+              typeInvolvesTypeVar(closureParamType->getResult(),
+                                  resultTypeVar)) {
+            increaseScore(SK_Fix);
+          }
+        } else if (auto *argTuple = getAsExpr<TupleExpr>(AE->getArg())) {
+          if (llvm::any_of(indices(argTuple->getElements()), [&](unsigned i) {
```

Comment:
Instead of doing ad hoc argument to parameter matching here let's try to use already recorded parameter bindings stored in `ConstraintSystem::argumentMatchingChoices` so both of these cases here could be unified.

Reply:
Ah awesome, I'll check that! thank you :)

Reply:
Yeah. I don't know how exactly we should handle this, maybe we shouldn't produce any kind of fix if contextual mismatch has already been detected for `T` but then there is no way to control ordering...

Reply:
Ah yeah, I remember with discuss something similar before in #34451, back then as far as I remember we were trying to detect those contextual mismatches at `diagnoseAmbiguityWithFixes` ... but in your opinion is this something we could explore more or there are holes in that approach? Just asking for curiosity :)

Reply:
Yes, maybe. It seems like we need a change of strategy here because locators just aren't cutting it in the situation when a single issue is manifested in multiple different locations. This makes me thing that the actual thing we need to diagnose here is something like `conflicting generic parameter 'T' is inferred as 'Int' from context and 'Void' in the body of the closure` or something like that...

Reply:
> This makes me thing that the actual thing we need to diagnose here is something like conflicting generic parameter 'T' is inferred as 'Int' from context and 'Void' in the body of the closure or something like that...

Interesting, that makes a lot of sense to me because in fact this is actually what is happening and also seems like something that would be possible to correctly detect in `diagnoseAmbiguityWithFixes`, I'll think about it more and see what we can do :)

Reply:
Sounds good!